abd-clam/src/cakes/knn/sieve_v1.rs

```diff
@@ -29,257 +27,165 @@ where
-    k: usize,
-    /// A vector of `Grains` which could still contain one of the k-nearest
-    /// neighbors. A `Grain` is a cluster, a distance, and a multiplicity.
-    grains: Vec<Grain<'a, T, U>>,
-    /// Whether hits contains the k-nearest neighbors.
-    is_refined: bool,
-    /// A priority queue of points which could be a nearest neighbor.
-    hits: Hits<U>,
-}
+    let mut grains = vec![Grain::new(c, d)];
+    let mut hits = Hits::new(k);
 
-impl<'a, T: Send + Sync + Copy, U: Number, D: Dataset<T, U>> SieveV1<'a, T, U, D> {
-    /// Creates a new instance of a `Sieve V1`.
-    pub fn new(tree: &'a Tree<T, U, D>, query: T, k: usize) -> Self {
-        Self {
-            tree,
-            query,
-            k,
-            grains: Vec::new(),
-            is_refined: false,
-            hits: Hits::default(),
-        }
-    }
+    loop {
+        // The threshold is the minimum distance, so far, which guarantees that
+        // the k nearest neighbors are within the threshold.
+        let threshold = Grain::partition(&mut grains, k).d;
 
-    /// One-time computation of `grains.`
-    pub fn initialize_grains(&mut self) {
-        let root = self.tree.root();
-        let distance = root.distance_to_instance(self.tree.data(), self.query);
-        let grain = Grain::new(root, distance + root.radius, root.cardinality);
-        self.grains = vec![grain];
-    }
+        // Remove hits which are outside the threshold.
+        hits.pop_until(threshold);
 
-    /// Returns whether search is complete, i.e., whether hits
-    /// contains the k-nearest neighbors.
-    const fn is_refined(&self) -> bool {
-        self.is_refined
-    }
+        // Partition the grains into outsiders and non-outsiders.
+        let (outsiders, non_outsiders) = grains.into_iter().partition::<Vec<_>, _>(|g| g.is_outside(threshold));
 
-    /// One iteration of the refinement step.
-    pub fn refine_step(&mut self) {
-        // Determines the index of the grain such that if we only retain that grain
-        // and every closer grain, we are guaranteed to have at least k points.
-        // Threshold is the distance of the furthest point in the grain.
-        let i = Grain::partition_kth(&mut self.grains, self.k);
-        let ith_grain = &self.grains[i];
-        let threshold = ith_grain.d;
-
-        // Filters hits by being outside the threshold.
-        while !self.hits.is_empty()
-            && self
-                .hits
-                .peek_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"))
-                .1
-                .number
-                > threshold
-        {
-            self.hits
-                .pop_max()
-                .unwrap_or_else(|| unreachable!("`hits` is non-empty"));
+        // If there are no outsiders then the search is complete.
+        if outsiders.is_empty() {
+            grains_to_hits(data, query, &mut hits, non_outsiders);
+            break;
```

Comment:
I trust you but I don't yet understand this so leaving a comment here for me to think this through. How is this not activating on the first iteration? The root cluster won't be an outsider so you'll have no outsiders so need to think through why this isn't just hitting on the root cluster